integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+            returnValue = new XAException(XAER_RMERR);
+        } else if (e instanceof XAException xaException) {
+            // Obviously if e is an XAException no conversion is necessary.
+            returnValue = xaException;
+        } else {
+            Throwable cause = e.getCause();
+            if (cause instanceof XAException xaException) {
+                // No matter what, if the cause was an XAException then it is canonical.
+                returnValue = xaException;
+            } else if (e instanceof IllegalTransitionException) {
+                // Any IllegalTransitionException is by definition an XA protocol problem.
+                returnValue = (XAException) new XAException(XAER_PROTO).initCause(e);
+            } else if (e instanceof SQLException sqlException) {
+                returnValue = this.exceptionConverter.convert(xaRoutine, sqlException);
+            } else if (cause instanceof SQLException sqlException) {
+                returnValue = this.exceptionConverter.convert(xaRoutine, sqlException);
+            } else {
+                returnValue = (XAException) new XAException(XAER_RMERR).initCause(e);
+            }
+        }
+        if (returnValue == null) {
+            returnValue = (XAException) new XAException(XAER_RMERR).initCause(e);
+        }
+        return returnValue;
+    }
+
+
+    /*
+     * Static methods.
+     */
+
+
+    private static void requireNonNullXid(Xid xid) throws XAException {
+        if (xid == null) {
+            throw (XAException) new XAException(XAER_INVAL).initCause(new NullPointerException("xid"));
+        }
+    }
+
+    // (Used via method reference only when an exceptionConverter was not supplied at construction time. This is the
+    // default implementation.)
+    private static XAException convert0(XARoutine xaRoutine, Exception e) {
+        if (e == null) {
+            return new XAException(XAER_RMERR);
+        } else if (e instanceof XAException xaException) {
+            return xaException;
+        } else {
+            Throwable cause = e.getCause();
+            if (cause instanceof XAException xaException) {
+                // No matter what, if the cause was an XAException then it is canonical.
+                return xaException;
+            } else {
+                SQLException sqlException;
+                if (e instanceof SQLException s) {
+                    sqlException = s;
+                } else if (cause instanceof SQLException s) {
+                    sqlException = s;
+                } else {
+                    sqlException = null;
+                }
+                if (sqlException != null) {
+                    String sqlState = sqlException.getSQLState();
+                    if (sqlState != null
+                        && (sqlState.startsWith("080")
+                            || sqlState.equalsIgnoreCase("08S01") // ("ess" not "five")
+                            || sqlState.equalsIgnoreCase("JZ006"))) {
```

Comment:
SQLStates are extremely specific: `08` indicates a class of error that designates some kind of a connection error, and the other specific codes cited here are common issues that places like StackOverflow reveal are problematic in the real world.